src/plugins/oidc/login.ts

```diff
@@ -57,70 +71,106 @@ const authorize = async (org: OrgData, scope: string) => {
-        : throwAssertNever(org.providerType);
+export const fetchOidcToken = async <T>(request: {
+  url: string;
+  init: RequestInit;
+}) => {
+  const { url, init } = request;
   const response = await fetch(url, init);
-
   if (!response.ok) {
     if (response.status === 400) {
       const data = await response.json();
       if (data.error === "authorization_pending") return undefined;
     }
     await validateResponse(response);
   }
-  return (await response.json()) as TokenResponse;
+  return (await response.json()) as T;
+};
+
+const oidcRequestBuilder = (org: OrgData, authorize: AuthorizeResponse) => {
+  if (org.providerType === undefined) {
+    throw "Login requires a configured provider type.";
+  }
+  validateProviderDomain(org);
+
+  return {
+    url:
+      org.providerType === "okta"
+        ? `https:${org.providerDomain}/oauth2/v1/token`
+        : org.providerType === "ping"
+          ? `https://${org.providerDomain}/${org.environmentId}/as/token`
+          : throwAssertNever(org.providerType),
+    init: {
+      method: "POST",
+      headers: OIDC_HEADERS,
+      body: urlEncode({
+        client_id: org.clientId,
+        device_code: authorize.device_code,
+        grant_type: DEVICE_GRANT_TYPE,
+      }),
+    },
+  };
 };
 
 /** Waits until user device authorization is complete
  *
  * Returns the OIDC token after completion.
  */
-const waitForActivation = async (
-  org: OrgData,
-  authorize: AuthorizeResponse
-) => {
-  const start = Date.now();
-  while (Date.now() - start <= authorize.expires_in * 1e3) {
-    const response = await fetchOidcToken(org, authorize);
-    if (!response) await sleep(authorize.interval * 1e3);
-    else return response;
-  }
-  throw "Expired awaiting in-browser authorization.";
-};
+export const waitForActivation =
+  <A, T>(authorize: A & { interval: number }) =>
+  async (
+    getExpiry: (authorize: A) => number, // Aws implementation differs from standard OIDC response, need function to extract expiry
+    fetchToken: (authorize: A) => Promise<T | undefined>
+  ) => {
+    const start = Date.now();
+    while (Date.now() - start <= getExpiry(authorize) * 1e3) {
+      const response = await fetchToken(authorize);
+      if (!response) await sleep(authorize.interval * 1e3);
+      else return response;
+    }
+    throw "Expired awaiting in-browser authorization.";
+  };
 
-/** Logs in to an Identity Provider via OIDC */
-export const oidcLogin = async (org: OrgData, scope: string) => {
-  if (org.providerType === undefined) {
-    throw "Login requires a configured provider type.";
-  }
-  const authorizeResponse = await authorize(org, scope);
-  print2(`Please use the opened browser window to continue your P0 login.
+export const oidcLoginSteps: (
+  org: OrgData,
+  scope: string
+) => OidcLoginSteps<AuthorizeResponse, TokenResponse> = (
+  org: OrgData,
+  scope: string
+) => ({
+  setup: async () => {
```

Comment:
nit: I would expect this to be called something like `validate` because it's not actually saving/storing any data. set up makes me think we're "installing" some data somehow.